### DISCUSSION TITLE
Better File Chunk | 更加强大的文件分块

### DISCUSSION THREAD

**Discussion by arvinxx** [lobehub/lobehub#3550] [General]
## 背景

在 RAG 中，只有将文件合理分块后，才能做好检索与查询，但是市面上文件类型是非常多的，目前一期只做了一部分的分块支持。

目前支持的分块类型：

纯文本类：

- [x] TXT
- [x] PDF
- [x] MD
- [x] MDX
- [x] Latex https://github.com/lobehub/lobe-chat/pull/3592
- [ ] lrc/lrcx 

代码类：

- [x] js (js,mjs,jsx)
- [x] TS(ts, tsx)
- [x] HTML
- [x] toml
- [x] yaml
- [x] json
- [x] sh
- [x] patch
- [x] log https://github.com/lobehub/lobe-chat/pull/4541

富文本类：
- [x] DOCX
- [x] PPTX
- [ ] DOC
- [ ] PPT
- [ ] epub https://github.com/lobehub/lobe-chat/pull/6317

表格类：
- [ ] XLSX (EXCEL)
- [ ] XLS
- [x] CSV https://github.com/lobehub/lobe-chat/pull/4123

音频类：
- [ ] MP3

视频类：
- [ ] MP4

如果有对文件类型的分块诉求，请在下面留言，并说明对此类文件的分块设想

**Comment by mujiannan** [comment]
excel可以说是知识库的刚需了，转成html然后借用html分块。

**Reply by arvinxx** [reply]
然而市面上并没有免费的 excel 分块方案。excel 转成 html 我也没找到现成方案。

要么就是用 Unstructed.io ，自己填这个 key。

如果要自己做的话比较麻烦，后续考虑成付费项吧。

**Reply by mujiannan** [reply]
可能引入下SpreadJS？不过这个包似乎太重了。
可以直接利用客户端运算资源转换，保持服务端轻量。
lobe-chat不带这个功能，用户倒是可以自行转换，Excel打开直接另存为就行，脚本随便写写也能自助批量转换，仅仅是用lobe-chat构建知识库时会不太爽利。

**Reply by arvinxx** [reply]
SpreadJS 是纯前端组件，而 excel 分块是在服务端做的。另外我看了下 SpreadJS 收费可不便宜： https://developer.mescius.com/spreadjs/pricing ，你确定个人能付得起 $999 就为了这个功能么…

现阶段可行的方案应该就只有 Unstructed.io 了

**Reply by wangping886** [reply]
dify支持了 TXT、 MARKDOWN、 PDF、 HTML、 **XLSX、 XLS**、 DOCX、 CSV、 MD、 HTML等 ；具体实现我还没看，lobechat是否可以参考一下？ 借助python实现的？ @arvinxx

**Comment by Kevin7Qi** [comment]
希望纯文本类型可以增加对tex文件的支持

**Reply by arvinxx** [reply]
tex 是什么格式？

**Reply by Kevin7Qi** [reply]
LaTeX文档

**Reply by arvinxx** [reply]
LateX 需要特殊处理吗？还是就单纯用 text 解析就好了？

**Reply by arvinxx** [reply]
找了下有 LateX 专门的 Splitter ， @Kevin7Qi 能麻烦发我一个 LateX 的示例文档吗？我要弄一个测试用例

**Reply by Kevin7Qi** [reply]
GitHub好像传不了tex文件，我放到Dropbox上了：[sample_latex_document.tex](https://www.dropbox.com/scl/fi/du3vxzgt7p71k4b4bmbvy/sample_latex_document.tex?rlkey=814ve4w5c94nxl78tpe42lw6r&st=4ppcf3pn&dl=0)

**Reply by arvinxx** [reply]
已支持： https://github.com/lobehub/lobe-chat/pull/3592

**Reply by Kevin7Qi** [reply]
感谢🙏

**Comment by arvinxx** [comment]
计划支持一波 Lrc/Lrcx，这样就可以做歌词文件的分析了

https://github.com/ddddxxx/LyricsX

**Reply by wajkzmn-droid** [reply]
Token中转站高速稳定，通用大模型，有测试，支持高并发，地址：https://aacool.cn/console

**Comment by baitwo02** [comment]
可以支持一下typst么？比较新的标记语言，和latex算是竞品。
胜在语法简单，环境配置也简单，比起latex友好不少。

**Reply by arvinxx** [reply]
能否给出一个示例？以及给出一下你期望的分块的模式？

**Reply by baitwo02** [reply]
示例：
```typst
== 函数极限与连续

=== 极限定义
$forall xi > 0$, $exists delta > 0$ ,当 $|x - x_(0)| < delta $时, $|f(x) - A| <= 0$ ,则 $limits(lim)_(x -> x_(0)) f(x) = A$

当$x -> 0$ 时, $ln(x + sqrt(1 + x^(2))) ~ x$, $ln(x + sqrt(1 + x^(2)))' = (1)/(sqrt(1 + x^(2)))$

=== 常见的大小关系比较

#block(fill:rgb("#f0f0f0"), inset: 10pt, radius:5pt, width: 100%)[
当 $x -> oo$ 时:

$ln^alpha x << x^beta << a^x$

$ln^alpha n << n^beta << a^n << n! << n^n$
]

#mblock[
$2/pi < sin x < x < tan x < 4/pi$
]

=== 泰勒公式

#block(fill:rgb("#f0f0f0"), inset: 10pt, radius:5pt, width: 100%)[
$sin(x) = x - (x^3)/3! + o(x^3)$

$cos x = 1 - x^2/2! + x^4/4! + o(x^(4))$

$arcsin x = x + x^3/3! + o(x^3)$

$tan x = x + x^3/3 + o(x^3)$

$arctan x = x - x^3/3 + o(x^(3))$

$ln(1 + x) = x - x^2/2 + x^3/3 + o(x^3)$

$e^x = 1 + x + x^2/2! + x^3/3! + o(x^3)$

$(1 + x)^alpha = 1 + alpha x + (alpha (alpha - 1) x^2)/2! + o(x^2)$

$ln(x + sqrt(1 + x^(2))) = x + (1)/(6) x^(3) + o(x^(3))$
]
```

至于分块模式我不是很了解，不知道怎样分比较好。

**Comment by dogeggo** [comment]
java和python有现有的分块方案吗,请求加上

**Reply by arvinxx** [reply]
最简单的就是当成一整个文件来切分，这样啥也不用做就直接能支持。但这样的问题是如果是超长代码块，可能会有太长的问题

**Comment by Forcome-Database** [comment]
请问Unstructed.io如何配置，项目里有集成 吗？

**Comment by githubdebugger** [comment]
Please add ePub support as well to the list of files supported? 

https://github.com/lobehub/lobe-chat/discussions/4594#discussioncomment-11133596

Most of the books are in ePub or PDF format, also seems like LLM's are better at reading ePub (due to less clutter than PDF's formatting).

You also could add Mobi support, but Mobi support is dying since AMZN (it was only AMZN who were supporting it) stopped supporting it and even they have moved onto ePub now.

Thanks

**Reply by arvinxx** [reply]
I will check it. Thanks for your advice~

**Comment by zmn817** [comment]
Unstructed.io 怎么使用？

**Comment by ginamber** [comment]
楼主好，我想到一个办法，可以把excel文件转换成YAML格式的文件，然后按一定的规则分块。

**Reply by ginamber** [reply]
贴一个站，可以把Excel转成YAML格式：https://tableconvert.com/zh-tw/excel-to-yaml

※分块想法，每两个“-xxx-”之间分一个块。
-
   商品ID: 605081056762
   商品名称: 商品A
   货号: 001
   商品状态: 当前在线
   商品访客数: 7,357.00 
   商品浏览量: 20,816
-
   商品ID: 606918572693
   商品名称: 商品B
   货号: 002
   商品状态: 当前在线
   商品访客数: 14,546
   商品浏览量: 31,381
-
   商品ID: xxxx
   商品名称: xxx
   货号: xxx
   商品状态: xxx
   商品访客数: xxx
   商品浏览量: xxx

**Comment by flyhunterl** [comment]
我在另外一个项目上是把excel表格转换成Markdown格式……就能向量了

**Comment by flyhunterl** [comment]
https://github.com/nanbingxyz/5ire    发现一个支持向量EXCEL的

**Reply by ginamber** [reply]
AnythingLLM也可以向量化xls，不知道为什么LobeChat弄不好，本来一直用Lobe的，因为没有Excel向量功能，我打算换平台了。

**Comment by mujiannan** [comment]
https://github.com/microsoft/markitdown

现在微软官方出了，office转markdown，py工具

**Comment by chexxor** [comment]
I have some XML WSDL files that describe my API.
When I upload it to my knowledgebase and attempt to chunk it, lobe-chat fails and says that XML is not a supported file type.

**Comment by dab512** [comment]
Please add transcription via whisper open ai and chunking for this transcribed text, and view via html5 media player with links to transcribed part by timing

**Comment by m0eak** [comment]
图片格式能支持一下吗，美术很需要

**Reply by Bet5521** [reply]
![image](https://github.com/user-attachments/assets/58d7c472-1320-4115-8a57-b5621e7e5133)
很简单一个txt文件上传，都提示向量化失败
哪位知道啥情况

**Reply by m0eak** [reply]
好像是你apikey无效

**Comment by QYQSDTC** [comment]
markdown 的分块能不能用 # 分块，即把标题和标题下的内容分做一个块，现在会把标题单独分一个块

**Comment by fqscfqj** [comment]
![image](https://github.com/user-attachments/assets/df4b1c94-231d-4c7b-aac8-6fbcdad0b558)
[printer.zip](https://github.com/user-attachments/files/18928392/printer.zip)

请求支持cfg文件

**Comment by 2247ds** [comment]
希望支持xml文件

**Comment by Failionaire** [comment]
![image](https://github.com/user-attachments/assets/54acd6db-19e7-4c09-97c4-285d21bc5987)
![image](https://github.com/user-attachments/assets/1ef03f40-3667-4818-9661-f0896695d528)
.GS files are essentially JavaScript.

**Comment by MRUIAW** [comment]
不太懂该如何分块, 但请求支持一下`.hpp`文件

**Comment by amDosion** [comment]
我自定义了变量：            
- DEFAULT_FILES_CONFIG=${embedding_model}
- embedding_model=ollama.embeddings/nomic-embed-text

现在能成功，但是我不知道是不是真的成功，怎么查看利用的是哪一个向量模型？

**Comment by FloydZhangHub** [comment]
是否可以支持.h文件的分块？目前是支持.cpp文件的分块的,.h文件是C++代码文件的头文件。

**Comment by FloydZhangHub** [comment]
像.docx/.doc/.pdf等富文本文件里如果存在图片，分块就很容易失败，请问你们也有这种情况吗？

**Comment by He-Xun** [comment]
图书常用的有四种：
pdf：分图片和文本版本的电子书，图片的不太好处理，需要ocr效果还不一定好
azw3：这个是版权格式，可以放弃支持
mobi：希望支持
epub：希望支持

还有个从视频里抽出来的字幕文件，常用有三种。
srt：希望优先支持
ssa：可以延后
sub：可以延后

有时候查询一个主题的材料时没有太多的文本材料，会直接把下载视频的字幕提取出来，然后让ai帮我找重点、解答问题，如果不需要我手工粘贴到对话框进行使用会方便很多，如果可以从多个视频的字幕，通过大模型的能力把一个主题的内容进行总结，找出视频的讲解的重点时间、内容。助益会非常大，希望能对字幕文件进行支持～

**Comment by HanadaLee** [comment]
能否支持一下c代码文件，我看cpp是支持的，但是c却不支持

**Comment by hi-fox** [comment]
Can you support python .pyw files? python .py files are already supported but not .pyw for some reason

**Comment by hi-fox** [comment]
Is this thread still monitored? I would like to re-request python .pyw files, as python .py files are already supported so the chunking method will be the exact same

**Comment by alokkumar353** [comment]
hii , @arvinxx i would like raise a issue that i uploaded the excel file and it is not being able to create chunk of that file to use it in the semantic search . 
<img width="1366" height="768" alt="image" src="https://github.com/user-attachments/assets/07d51c38-9bf5-4f8d-97bd-85ea7fccb29f" />

**Comment by Chincherry93** [comment]
matlab的程序文件.m，建议增加读写的能力

**Comment by liyu9202** [comment]
建议添加qt软件qml文件支持，增加读写能力
